### PR TITLE
Switch to Server Builder

### PR DESCRIPTION
**Pull request by ChristopherDavenport**

### CODE REVIEW

File: modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import org.http4s._
+import org.http4s.server.Server
+import scala.concurrent.duration._
+import java.net.InetSocketAddress
+import java.util.concurrent.Executors
+import java.nio.channels.AsynchronousChannelGroup
+
+
+final class EmberServerBuilder[F[_]: Concurrent: Timer: ContextShift] private (
+  private val host: String,
```
Comment (rossabaker) [inline modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala]: The constructor needs to be private, but there's not much harm in exposing read access to the values that need write access.  I wouldn't typically make these vals private.

File: modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  private val port: Int,
+  private val httpApp: HttpApp[F],
+  private val agR: Resource[F, AsynchronousChannelGroup],
```
Comment (rossabaker) [inline modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala]: This is awkward.  I'm not fond of passing `Resource`s or effects as config, but it makes it hard to create a builder without it.  A way around it is to make it a required field in the companion `apply`, but that has kind of crap ergonomics.

File: modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  private val onError: Throwable => Response[F],
+  private val onWriteFailure : Option[(Option[Request[F]], Response[F], Throwable) => F[Unit]],
```
Comment (rossabaker) [inline modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala]: This type makes my head hurt.  The outer option in particular.  Could it not default to `Function.const(F.unit)`?

File: modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    requestHeaderReceiveTimeout = requestHeaderReceiveTimeout
+  )
+
+  def withHost(host: String) = copy(host = host)
+  def withPort(port: Int) = copy(port = port)
+  def withHttpApp(httpApp: HttpApp[F]) = copy(httpApp = httpApp)
+  def withAsynchronousChannelGroup(acg: AsynchronousChannelGroup) =
+    copy(agR = acg.pure[Resource[F, ?]])
+  def withOnError(onError: Throwable => Response[F]) = copy(onError = onError)
+  def withOnWriteFailure(onWriteFailure: Option[(Option[Request[F]], Response[F], Throwable) => F[Unit]]) =
```
Comment (rossabaker) [inline modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala]: Our other builders that take an `Option` usually call the field `Option`, and then have a `withOnWriteFailure` and `withoutWriteFailure`.  Those names are awkward, which I think you can work around without the outer `Option`.

File: modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala (hunk 1 above)
Comment (ChristopherDavenport) [reply modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala]: Exposed now

File: modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala (hunk 3 above)
Comment (ChristopherDavenport) [reply modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala]: Didn't like `Function.const`, but I could lose the Option.

File: modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala (hunk 4 above)
Comment (ChristopherDavenport) [reply modules/server/src/main/scala/io/chrisdavenport/ember/server/EmberServerBuilder.scala]: Dropped Option so we should be cleaner now.

**Pull request merged**